cpu/em_quad_level_frequency_transitions.py

```diff
@@ -0,0 +1,163 @@
+        2. Cpupower tool is installed.
+        """
+
+        if 'ppc' not in distro.detect().arch:
+            self.cancel("Processor is not ppc64")
+        if not os.path.exists('/sys/devices/system/cpu/cpu0/cpufreq'):
+            self.cancel('CPUFREQ is supported only on Power NV')
+
+        smm = SoftwareManager()
+        detected_distro = distro.detect()
+        self.threshold = int(self.params.get("threshold", default=300000))
+        if 'Ubuntu' in detected_distro.name:
+            deps = ['linux-tools-common', 'linux-tools-%s'
+                    % platform.uname()[2]]
+        elif detected_distro.name == "SuSE":
+            deps = ['cpupower']
+        else:
+            deps = ['kernel-tools']
+        for package in deps:
+            if not smm.check_installed(package) and not smm.install(package):
+                self.cancel('%s is needed for the test to be run' % package)
+
+        self.cpus = cpu.total_cpus_count()
+        self.cpu_num = 0
+        self.max_freq = 0
+
+    def test(self):
+        """
+        1) Change governor to userspace on all CPUs.
+        2) For each core in the quad, pick a random frequency
+        and set cpu with that frequency.
+        3) Validate that the cpuinfo_cur_freq on any core in
+        the code is set to max(set of frequencies)
+        4) Run the perf command on each cpu to validate frequencies
+        independently.
+        """
+
+        self.quad_dict = {}
+        self.max_freq_dict = {}
+        self.quad_to_cpu_mapping()
+        output = process.run("cpupower frequency-set -g userspace", shell=True)
+        cur_governor = self.cpu_freq_path('scaling_governor')
+        if 'userspace' == cur_governor and output.exit_status == 0:
+            self.log.info("%s governor set successfully" % cur_governor)
+        else:
+            self.cancel("Unable to set the userspace governor")
+        for chip in self.quad_dict:
+            for quad in self.quad_dict[chip]:
+                for self.cpu_num in self.quad_dict[chip][quad]:
+                    process.run("cpupower -c %s frequency-set -f %s"
+                                % (self.cpu_num, self.get_random_freq()))
+                    time.sleep(1)
+                    freq_set = self.cpu_freq_path('cpuinfo_cur_freq')
+                    if self.max_freq < freq_set:
+                        self.max_freq = freq_set
+                if chip not in self.max_freq_dict.keys():
+                    self.max_freq_dict[chip] = {}
+                if quad not in self.max_freq_dict[chip].keys():
+                    self.max_freq_dict[chip][quad] = self.max_freq
+                    self.log.info("Maximum frequency set:%s quad:"
+                                  "%s" % (self.max_freq, quad))
+                self.max_freq = 0
+        for chip in self.quad_dict:
+            for quad in self.quad_dict[chip]:
+                for cpu in self.quad_dict[chip][quad]:
```

Comment:
The values in dict can be got better with some of the below:
```
>>> dir({})
['__class__', '__cmp__', '__contains__', '__delattr__', '__delitem__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__iter__', '__le__', '__len__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setitem__', '__sizeof__', '__str__', '__subclasshook__', 'clear', 'copy', 'fromkeys', 'get', 'has_key', 'items', 'iteritems', 'iterkeys', 'itervalues', 'keys', 'pop', 'popitem', 'setdefault', 'update', 'values', 'viewitems', 'viewkeys', 'viewvalues']
```

To start with: 'items', 'iteritems', 'iterkeys', 'itervalues'

Reply:
let me try this , good suggestion

Reply:
@narasimhan-v can this taken as enhancement ?

Reply:
@PraveenPenguin Can we get a job.log please ?